File: 01.剑指offer/09.用两个栈实现队列/cqueue.cpp
//
// Created by lifan on 2022/5/22.
//
#include <stack>
#include <iostream>

class CQueue {
 private:
  std::stack<int> instack, outstack;

 public:
  CQueue() {
  }

  void appendTail(int value) {
      instack.push(value);
  }

  int deleteHead() {
      if (!outstack.empty()) {
          int val = outstack.top();
          outstack.pop();
          return val;
      }

      if (instack.empty()) {
          return -1;
      }

      while (!instack.empty()) {
          outstack.push(instack.top());
          instack.pop();
      }

      int val = outstack.top();
      outstack.pop();

      return val;
  }
};

int main(int argc, char *argv[]) {
    CQueue *queue = new CQueue();

    queue->appendTail(1);
    queue->appendTail(2);
    queue->appendTail(3);

    std::cout << queue->deleteHead() << std::endl;
    std::cout << queue->deleteHead() << std::endl;
    std::cout << queue->deleteHead() << std::endl;
    std::cout << queue->deleteHead() << std::endl;

    return 0;
}

File: 01.剑指offer/09.用两个栈实现队列/cqueue.py
class CQueue:
    def __init__(self):
        self.stack_in = list()
        self.stack_out = list()

    def appendTail(self, value: int) -> None:
        self.stack_in.insert(0, value)

    def deleteHead(self) -> int:
        if len(self.stack_out) == 0:
            for v in self.stack_in:
                self.stack_out.insert(0, v)
                self.stack_in.pop(0)

        if len(self.stack_out) == 0:
            return -1

        return self.stack_out.pop(0)


if __name__ == '__main__':
    q = CQueue()
    q.appendTail(1)
    q.appendTail(2)
    q.appendTail(3)
    print(q.deleteHead())
    print(q.deleteHead())
    print(q.deleteHead())
    print(q.deleteHead())
